### PR TITLE
Use EndpointReference instead of AllocatedEndpoints

### PR DESCRIPTION
**Pull request by davidfowl**
- This unifies most of the callers way of property looking up named endpoints and fixed case sensivity issues.
- Added a GetEndpoints method which returns all allocated endpoints on a resource.

Fixes #2521 
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2700)

### CODE REVIEW

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs [hunk 1]
```diff
@@ -227,40 +225,6 @@ public void WriteInputs(IResource resource)
         }
     }
 
-    /// <summary>
-    /// TODO: Doc Comments
-    /// </summary>
-    /// <param name="resource"></param>
-    public void WriteServiceDiscoveryEnvironmentVariables(IResource resource)
```
Comment (davidfowl) [inline src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:234]: Oops, I meant to put this back 😄

File: src/Aspire.Hosting/Extensions/ResourceExtensions.cs [hunk 2]
```diff
@@ -93,11 +93,26 @@ public static bool TryGetEndpoints(this IResource resource, [NotNullWhen(true)]
     /// <param name="resource">The resource to get the allocated endpoints for.</param>
     /// <param name="allocatedEndPoints">When this method returns, contains the allocated endpoints for the specified resource, if they exist; otherwise, <c>null</c>.</param>
     /// <returns><c>true</c> if the allocated endpoints were successfully retrieved; otherwise, <c>false</c>.</returns>
-    public static bool TryGetAllocatedEndPoints(this IResource resource, [NotNullWhen(true)] out IEnumerable<AllocatedEndpointAnnotation>? allocatedEndPoints)
+    private static bool TryGetAllocatedEndPoints(this IResource resource, [NotNullWhen(true)] out IEnumerable<AllocatedEndpointAnnotation>? allocatedEndPoints)
```
Comment (davidfowl) [inline src/Aspire.Hosting/Extensions/ResourceExtensions.cs:96]: Breaking change

**Comment by mitchdenny** [comment]
This is a nice change.

**Review by mitchdenny — APPROVED** [review]

**Pull request merged**